File: 02_Core_Project/council_navigator/client/src/lib/librarianInputGate.ts
/** Deterministic grammar gate for Librarian queries. */

export const GATE_VERSION = "grammar-v2";
export const QUERY_CHAR_CAP = 200;

export const GATE_MESSAGES = {
  not_a_string: "Write one question in the box.",
  empty: "Write a question before sending.",
  too_long: "Keep it under 200 characters — one focused question.",
  control_characters:
    "Use spaces between words instead of tabs, line breaks, or other control characters.",
  non_ascii:
    "Use plain English letters only — write names without accents and spell numbers out as words.",
  digits: "Spell numbers out as words — write 'sixteen million', not '16M'.",
  no_terminal_question_mark: "End your question with one question mark.",
  multiple_question_marks: "Use one question mark, at the end.",
  multiple_sentences:
    "Ask one focused question without periods, exclamation marks, semicolons, or colons.",
  no_words: "Write at least one word before the question mark.",
  bad_word:
    "Use words made from letters, apostrophes, or hyphens; a comma sits right after a word — like 'approved, and' — not {token}.",
} as const;

export type GateReasonCode = keyof typeof GATE_MESSAGES;

export interface LibrarianGateResult {
  ok: boolean;
  canonicalQuery?: string;
  reasonCode?: GateReasonCode;
  message?: string;
}

const WORD_RE = /^[A-Za-z]+(?:['-][A-Za-z]+)*'?,?$/;

function reject(
  reasonCode: GateReasonCode,
  token?: string
): LibrarianGateResult {
  const message =
    token === undefined
      ? GATE_MESSAGES[reasonCode]
      : GATE_MESSAGES[reasonCode].replace("{token}", JSON.stringify(token));
  return { ok: false, reasonCode, message };
}

export function validateLibrarianQuery(
  raw: string | null | undefined
): LibrarianGateResult {
  if (typeof raw !== "string") {
    return reject("not_a_string");
  }

  if (Array.from(raw).length > QUERY_CHAR_CAP) {
    return reject("too_long");
  }

  if (
    Array.from(raw).some(char => {
      const codePoint = char.codePointAt(0)!;
      return codePoint < 0x20 || codePoint === 0x7f;
    })
  ) {
    return reject("control_characters");
  }

  if (Array.from(raw).some(char => char.codePointAt(0)! > 0x7e)) {
    return reject("non_ascii");
  }

  if (/[0-9]/.test(raw)) {
    return reject("digits");
  }

  const canonical = raw.trim().replace(/ +/g, " ");
  if (Array.from(canonical).length > QUERY_CHAR_CAP) {
    throw new Error("Canonical Librarian query exceeded its raw length");
  }
  if (!canonical) {
    return reject("empty");
  }

  if (!canonical.endsWith("?")) {
    return reject("no_terminal_question_mark");
  }

  if (canonical.split("?").length - 1 !== 1) {
    return reject("multiple_question_marks");
  }

  if (/[.!;:]/.test(canonical)) {
    return reject("multiple_sentences");
  }

  const body = canonical.slice(0, -1).trim();
  if (!body) {
    return reject("no_words");
  }

  const tokens = body.split(" ");
  for (let index = 0; index < tokens.length; index += 1) {
    const token = tokens[index];
    if (!WORD_RE.test(token)) {
      return reject("bad_word", token);
    }
    if (index === tokens.length - 1 && token.endsWith(",")) {
      return reject("bad_word", token);
    }
  }

  return { ok: true, canonicalQuery: canonical };
}

File: 02_Core_Project/council_navigator/parsers/librarian_input_gate.py
"""Deterministic grammar-v2 gate for Librarian queries."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Optional


GATE_VERSION = "grammar-v2"
QUERY_CHAR_CAP = 200

MESSAGES: dict[str, str] = {
    "not_a_string": "Write one question in the box.",
    "empty": "Write a question before sending.",
    "too_long": "Keep it under 200 characters — one focused question.",
    "control_characters": (
        "Use spaces between words instead of tabs, line breaks, or other "
        "control characters."
    ),
    "non_ascii": (
        "Use plain English letters only — write names without accents and "
        "spell numbers out as words."
    ),
    "digits": (
        "Spell numbers out as words — write 'sixteen million', not '16M'."
    ),
    "no_terminal_question_mark": (
        "End your question with one question mark."
    ),
    "multiple_question_marks": (
        "Use one question mark, at the end."
    ),
    "multiple_sentences": (
        "Ask one focused question without periods, exclamation marks, "
        "semicolons, or colons."
    ),
    "no_words": "Write at least one word before the question mark.",
    "bad_word": (
        "Use words made from letters, apostrophes, or hyphens; a comma sits "
        "right after a word — like 'approved, and' — not {token}."
    ),
}

_WORD_RE = re.compile(r"[A-Za-z]+(?:['-][A-Za-z]+)*'?,?")


@dataclass(frozen=True)
class GateResult:
    ok: bool
    canonical_query: Optional[str]
    reason_code: Optional[str]
    message: Optional[str]


def _reject(reason_code: str, *, token: str | None = None) -> GateResult:
    message = MESSAGES[reason_code]
    if token is not None:
        message = message.format(token=json.dumps(token))
    return GateResult(
        ok=False,
        canonical_query=None,
        reason_code=reason_code,
        message=message,
    )


def validate_librarian_query(raw: object) -> GateResult:
    """Validate and canonicalize one Librarian question."""

    if not isinstance(raw, str):
        return _reject("not_a_string")

    if len(raw) > QUERY_CHAR_CAP:
        return _reject("too_long")

    if any(ord(char) < 0x20 or ord(char) == 0x7F for char in raw):
        return _reject("control_characters")

    if any(ord(char) > 0x7E for char in raw):
        return _reject("non_ascii")

    if any("0" <= char <= "9" for char in raw):
        return _reject("digits")

    canonical = " ".join(raw.strip().split())
    assert len(canonical) <= QUERY_CHAR_CAP
    if not canonical:
        return _reject("empty")

    if not canonical.endswith("?"):
        return _reject("no_terminal_question_mark")

    if canonical.count("?") != 1:
        return _reject("multiple_question_marks")

    if any(mark in canonical for mark in ".!;:"):
        return _reject("multiple_sentences")

    body = canonical[:-1].strip()
    if not body:
        return _reject("no_words")

    tokens = body.split(" ")
    for index, token in enumerate(tokens):
        if _WORD_RE.fullmatch(token) is None:
            return _reject("bad_word", token=token)
        if index == len(tokens) - 1 and token.endswith(","):
            return _reject("bad_word", token=token)

    return GateResult(
        ok=True,
        canonical_query=canonical,
        reason_code=None,
        message=None,
    )
